src/app/functions/server/getRelevantGuidedSteps.ts

```diff
@@ -0,0 +1,241 @@
+    id: "LeakedPasswordsSecurityQuestion",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/phone",
+    id: "SecurityTipsPhone",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/email",
+    id: "SecurityTipsEmail",
+  },
+  {
+    href: "/redesign/user/dashboard/fix/security-recommendations/ip",
+    id: "SecurityTipsIp",
+  },
+  {
+    href: "/redesign/user/dashboard",
+    id: "Done",
+  },
+] as const satisfies ReadonlyArray<{ href: string; id: string }>;
+
+export type StepLink = (typeof stepLinks)[number];
+
+export type OutputData = {
+  current: StepLink | null;
+  skipTarget: StepLink | null;
+};
+
+export function getRelevantGuidedSteps(
+  data: InputData,
+  stepsToSkip: Array<StepLink["id"]> = []
+): OutputData {
+  const current = getNextGuidedStep(data, stepsToSkip);
+
+  // There should be a relevant next step for every user (even if it's just
+  // going back to the dashbord), so we can't hit this line in tests (and
+  // shouldn't be able to in production either):
+  /* c8 ignore next 6 */
+  if (current === null) {
+    return {
+      current: null,
+      skipTarget: null,
+    };
+  }
+
+  const nextStepsToSkip = [...stepsToSkip];
+  nextStepsToSkip.push(current.id);
+  return {
+    current: current,
+    skipTarget: getNextGuidedStep(data, nextStepsToSkip),
+  };
+}
+
+function getNextGuidedStep(
+  data: InputData,
+  stepsToSkip: Array<StepLink["id"]> = []
+): StepLink | null {
+  // Resisting the urge to add a state machine... ^.^
+  const nextStep = stepLinks.find((stepLink) => {
+    return (
+      !stepsToSkip.includes(stepLink.id) &&
+      isEligibleFor(data, stepLink.id) &&
+      !hasCompleted(data, stepLink.id)
+    );
+  });
+
+  return nextStep ?? null;
+}
+
+function isEligibleFor(data: InputData, stepId: StepLink["id"]): boolean {
+  if (stepId === "StartScan") {
+    return data.countryCode === "us";
+  }
+
+  if (stepId === "ScanResult") {
+    return data.countryCode === "us";
+  }
+
+  if (stepId === "HighRiskSsn") {
+    // Our social security number-related mitigations aren't possible outside of the US:
+    return data.countryCode === "us";
+  }
+
+  if (
+    ["HighRiskCreditCard", "HighRiskBankAccount", "HighRiskPin"].includes(
+      stepId
+    )
+  ) {
+    // Anyone can view/resolve their high risk data breaches
+    return true;
+  }
+
+  if (
+    ["LeakedPasswordsPassword", "LeakedPasswordsSecurityQuestion"].includes(
+      stepId
+    )
+  ) {
+    // Anyone can view/resolve their high risk data breaches
+    return true;
+  }
+
+  if (
+    ["SecurityTipsPhone", "SecurityTipsEmail", "SecurityTipsIp"].includes(
+      stepId
+    )
+  ) {
+    // Anyone can view security tips
+    return true;
+  }
+
+  if (stepId === "Done") {
+    return true;
+    // All steps should have been covered by the above conditions:
+    /* c8 ignore next 4 */
+  }
+
+  return false as never;
```

Comment:
Question: If we would never return anything here this is basically dead code? In case this is here to catch any logic errors when tweaking the above logic in the future: Would a warning log with a message be more helpful to have?

Reply:
Yeah, this should be dead code. I was actually hoping that this would trigger TS errors if I forgot to handle a particular step above, but that doesn't seem to happen :/ We could add a log, but I think this would then either never log, or suddenly fill up our logs entirely if we'd forgotten something.